### PR TITLE
LogContext: Fix scroll position in upper context group

### LINKED ISSUE #56122: Show context should open centered on the lines closest to the line you're showing context for
Actually looking closer maybe this is a regression? This appears that it should have been fixed here:
https://github.com/grafana/grafana/issues/50268

-------

When you open the show context modal, Grafana fetches the next 10 lines that come after my selected line. 

The modal that pops up aligns itself to show the lines that are furthest away (furthest in the future) from my selected line. 

This is confusing for me as a user, since my workflow is to start on the line I'm interested in (the one I asked to show context for) and then scroll gradually further back in time, or further forward in time. 

To better fit this workflow, I would expect the modal to pop up centered on the lines that are nearest to the one I'm looking at the context for. So I should see the lines that are immediately in the future and immediately in the past and then be able to gradually scroll away from those. 

Here's what it looks like today. If you look at the red arrow, you'll see that there's a bunch of lines that are closer in time to the line I'm looking at the context for that are being truncated. The view is set with the line pointed to by the light blue arrow at the top of the screen
Screen Shot 2022-09-30 at 12 42 29 PM

Here's what I'd expect -- that the view pops up with the scroll set so I see the line that's closest to the line that I'm looking at the context for. See the red curly brace. 
Screen Shot 2022-09-30 at 12 43 32 PM

### PR DESCRIPTION
**Pull request by svennergr**
**What this PR does / why we need it**:
The scroll position of the upper context group does currently not keep the correct position. It should initially scroll to the bottom and also keep at the current scroll position.

Initially this was fixed in #50600 however that fix was only working if less or equal than 10 rows were added. However with the change of the IDs of logrows in https://github.com/grafana/grafana/pull/53932 we accidentally added more than 10 rows to the upper context. The PR fixes that issue by adding the correct refid to the comparison and adds a check to only add the requested amount of loglines.

Furthermore the PR improves the handling of the scroll position, such that the position will be maintained also if further lines are loaded.

**Which issue(s) this PR fixes**:

Fixes #56122

**Special notes for your reviewer**:

https://user-images.githubusercontent.com/8092184/194031849-a68dd78d-484d-4f11-90f6-76fb98e7f2c1.mov

### CODE REVIEW

File: public/app/features/logs/components/LogRowContext.tsx [hunk 1]
```diff
@@ -1,11 +1,12 @@
...
+import React, { useEffect, useLayoutEffect, useRef, useState } from 'react';
+import usePrevious from 'react-use/lib/usePrevious';
 
-import { GrafanaTheme2, DataQueryError, LogRowModel, textUtil, LogsSortOrder } from '@grafana/data';
-import { useStyles2, Alert, ClickOutsideWrapper, CustomScrollbar, List, Button } from '@grafana/ui';
+import { DataQueryError, GrafanaTheme2, LogRowModel, LogsSortOrder, textUtil } from '@grafana/data';
+import { Alert, Button, ClickOutsideWrapper, CustomScrollbar, List, useStyles2 } from '@grafana/ui';
 
 import { LogMessageAnsi } from './LogMessageAnsi';
-import { LogRowContextRows, LogRowContextQueryErrors, HasMoreContextRows } from './LogRowContextProvider';
```
Comment (svennergr) [inline public/app/features/logs/components/LogRowContext.tsx:8]: Besides the addition of `usePrevious` those are just the result of an `organize imports` action.

File: public/app/features/logs/components/LogRowContextProvider.tsx [hunk 2]
```diff
@@ -93,8 +93,14 @@ export const getRowContexts = async (
...
-          //counts.
-          if (idField.values.get(fieldIndex) === row.uid) {
+          // counts.
+          // We add the refId of the original dataFrame, since uids contain those now and the uids in the context-dataframe
+          // do not have a refid.
+          // To be compatible to other datasources we check for both IDs with and without refId.
+          if (
+            idField.values.get(fieldIndex) === row.uid ||
+            `${idField.values.get(fieldIndex)}${row.dataFrame.refId}` === row.uid
+          ) {
```
Comment (svennergr) [inline public/app/features/logs/components/LogRowContextProvider.tsx:103]: @gabor I am unsure about this. Should we check for both or just 
```
`${idField.values.get(fieldIndex)}${row.dataFrame.refId}` === row.uid
```

Comment (svennergr) [reply public/app/features/logs/components/LogRowContextProvider.tsx:103]: Nevermind. I made an other approach by adding the original refid to the context-query.

Comment (svennergr) [reply public/app/features/logs/components/LogRowContextProvider.tsx]: https://github.com/grafana/grafana/pull/56370/files#diff-2e4d3a3dbd431f921794de33720c6fc3ac7a8a2f6dbde31424e487080bb572c1R593

**Review by matyax — APPROVED** [review]
Looks good!

File: public/app/features/logs/components/LogRowContext.tsx [hunk 3]
```diff
@@ -155,15 +156,48 @@ export const LogRowContextGroup: React.FunctionComponent<LogRowContextGroupProps
...
   const [scrollTop, setScrollTop] = useState(0);
+  const [scrollHeight, setScrollHeight] = useState(0);
+
   const listContainerRef = useRef<HTMLDivElement>(null);
+  const prevRows = usePrevious(rows);
+  const prevScrollTop = usePrevious(scrollTop);
+  const prevScrollHeight = usePrevious(scrollHeight);
 
+  /**
+   * This hook is responsible to keep the right scroll position of the top
```
[the remark this answers is not in the document]
Comment (svennergr) [reply public/app/features/logs/components/LogRowContext.tsx]: Thanks, good suggestion!

File: public/app/features/logs/components/LogRowContext.tsx [hunk 4]
```diff
@@ -1,11 +1,12 @@
...
+import usePrevious from 'react-use/lib/usePrevious';
 
-import { GrafanaTheme2, DataQueryError, LogRowModel, textUtil, LogsSortOrder } from '@grafana/data';
-import { useStyles2, Alert, ClickOutsideWrapper, CustomScrollbar, List, Button } from '@grafana/ui';
+import { DataQueryError, GrafanaTheme2, LogRowModel, LogsSortOrder, textUtil } from '@grafana/data';
+import { Alert, Button, ClickOutsideWrapper, CustomScrollbar, List, useStyles2 } from '@grafana/ui';
 
 import { LogMessageAnsi } from './LogMessageAnsi';
-import { LogRowContextRows, LogRowContextQueryErrors, HasMoreContextRows } from './LogRowContextProvider';
+import { HasMoreContextRows, LogRowContextQueryErrors, LogRowContextRows } from './LogRowContextProvider';
```
[the remark this answers is not in the document]
Comment (svennergr) [reply public/app/features/logs/components/LogRowContext.tsx:10]: Definitely. We should refactor the whole LogRowContext in Q4 anyways, so those components would move (hopefully) anyways.

**Pull request merged**